### ISSUE TITLE
backends modularity breaks x11 session

### ISSUE THREAD

**Issue by 0323pin** [leftwm/leftwm#1245]
### Describe the bug

Building from the latest 765a1aeb175e5ad2027982a78546b29c8cba125b breaks my X-session, started as https://github.com/leftwm/leftwm/discussions/1244

### Expected behavior (optional)

Be able to run a X-session from the latest HEAD.

### Steps to reproduce (optional)

Build from 765a1aeb175e5ad2027982a78546b29c8cba125b and `startx`.

### Relevant log output (optional)

```text
NA, the session ends immediately and kicks me back to tty.
```

### OS / Distro

Void Linux - musl - x86_64

### Additional System Information (optional)

```text
Building with "no-default-features", features "sys-log, lefthk, xlib" or with, features "sys-log, lefthk, xlib, x11rb" result in not being able to launch a X-session.

LeftWM is launched from `.xinitrc` using `exec dbus-launch --sh-syntax --exit-with-session leftwm` as I've always been doing, for years.

Reverting back to 671fab081d675b25539cd6e241bc3d8063847cfb allows me to have a stable X-session.
```

### LeftWM Check

```text
NA, I cannot run this as the session ends immediately
```

**Comment by 0323pin** [comment]
Seems like I need to re-build on Linux and see what's wrong.

Building on NetBSD, I get only a build warning,
```
   Compiling x11rb-display-server v0.1.0 (/usr/pkgsrc/wip/leftwm/work/leftwm-765a1aeb175e5ad2027982a78546b29c8cba125b/display-servers/x11rb-display-server)
warning: unused import: `std::backtrace::BacktraceStatus`
 --> display-servers/x11rb-display-server/src/event_translate.rs:1:5
  |
1 | use std::backtrace::BacktraceStatus;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
```

and a working X-session when launching with `exec dbus-run-session leftwm` :thinking: 

![2024-04-04-111606_1366x768_scrot](https://github.com/leftwm/leftwm/assets/90570748/c6e24aa8-5498-4fac-b802-8274faa07155)

**Comment by Eskaan** [comment]
@0323pin The pr should've technically been a development change / new feature as it defaults to the XLib display server. I recommend looking at changes that have been made outside of `display-servers/x11rb-display-server`.

I will try to investigate this myself (using `startx` from tty with `exec leftwm` in ~/.xinitrc on Arch Linux systemd).

**Comment by Eskaan** [comment]
@0323pin I have added a log-config setting with this pr (d5ca210), I recommend to take a look at that specifically since sys-log should be the main intended difference.

You should be able to run `leftwm-check` as long as leftwm compiles, it's completetly separate from everything else.

**Comment by 0323pin** [comment]
What I don't get yet, is why its working on NetBSD and not on Void linux. Hence, I suspect a missing dependency. `x11` is part of the base system on NetBSD while, on Void I'm running `xorg-minimal`. I've been trying to find what `x11rb` needs. Some `xcb-util` package or, `xorg-proto` or, something else.

I don't know when I'll have time to check this again. I don't use my Linux system daily but, I'll read through your links.
Thanks.

**Comment by 0323pin** [comment]
> You should be able to run `leftwm-check` as long as leftwm compiles, it's completetly separate from everything else.

True but, I'll be on tty.
Another difference is that on NetBSD I've built with `file-log` instead of `sys-log` like on Void. Both builds are fine but, the X-session works only on NetBSD for now.

**Comment by Syudagye** [comment]
> ### Relevant log output (optional)
> 
> ```
> NA, the session ends immediately and kicks me back to tty.
> ```

Do you have any logs from xinit/startx that could help ? I believe you can get them by redirecting stdout or stderr when running `startx` to a file ( i don't recall which one holds the sub-programs stdout)

**Comment by 0323pin** [comment]
Ok, it seems that this is really an issue when building with the `sys-log` feature. I can cross-check this on NetBSD tomorrow or, Saturday but, I didn't want to keep you guys waiting. Thanks @Eskaan for the little hint.

I've changed my usual build on Linux using `sys-log` to use `file-log` and I have **not** changed my config file.
Building with `cargo build --release --no-default-features --features lefthk,file-log,xlib,x11rb` on Void - musl - x86_64 yields a working X-session, just like on NetBSD :smiley: 

![2024-04-04-222748_1920x1080_scrot](https://github.com/leftwm/leftwm/assets/90570748/b7ce040d-8f6a-4875-a62e-8e093025afcf)

Now, something's wrong with the `sys-log` build. I can check this again pulling from @Eskaan branch, https://github.com/leftwm/leftwm/commit/d5ca2102df1abe7d010e145aaffa53d3f9cf4e88 but, I won't be able to do it this week, maybe @hertg if he has time?

At least, I'm now running on this commit. Still, need to test the backend switch but, at least my stable machine is working again. Thanks!

Keeping this issue open until the `sys-log` build works.

**Comment by mautamu** [comment]
Thanks for opening this @0323pin! In that case, are we good to go ahead and merge https://github.com/leftwm/leftwm/pull/1243?

**Comment by 0323pin** [comment]
> are we good to go ahead and merge #1243?

I didn't find the time to test this. I might be able to do it during the weekend.

**Comment by mautamu** [comment]
> I didn't find the time to test this. I might be able to do it during the weekend.

I don't think this issue is related to that fix

**Comment by 0323pin** [comment]
@mautamu Ahh ... sorry, I thought you were refering to https://github.com/leftwm/leftwm/commit/d5ca2102df1abe7d010e145aaffa53d3f9cf4e88, my bad

**Comment by Eskaan** [comment]
@0323pin
Corresponding line:
https://github.com/leftwm/leftwm/blob/765a1aeb175e5ad2027982a78546b29c8cba125b/leftwm/src/utils/log/sys.rs#L20

According to documentation, this should only fail if there already is a syslog with that identifier. ([docs ref](https://docs.rs/syslog-tracing/latest/syslog_tracing/struct.Syslog.html#method.new))
That makes it quite obvious to me why it breaks: The boot-logger I added isn't correctly dropped before creating the new one.

Specifically, it looks like we have to move this line up above the initialization of the new logger:
https://github.com/leftwm/leftwm/blob/765a1aeb175e5ad2027982a78546b29c8cba125b/leftwm/src/bin/leftwm-worker.rs#L43

I designed it this way so that we could initialize and set the global logger before dropping the boot-logger (no log-free time), but it seems like this won't work with sys-log.
